### DISCUSSION TITLE
I want to know the difference between the entity class in model package under the `data module` and the entity class under the `model module`

### DISCUSSION THREAD

**Discussion by est7** [android/nowinandroid#728] [Q&A] [answered]
I am confused about why there are so many entity classes stored separately, such as core.database.model.NewsResourceEntity, core.network.model.NetworkNewsResource, core.model.data.NewsResource, and core.model.data.FollowableTopic.

I want to know how to determine which module these entity classes belong to, where the asEntity function should be created, what should be stored under core.data.model, and which entity UI layer should depend on.

I am very grateful for answering my doubts.

**Comment by tunjid** [comment, marked as answer]
Hi there!

## Breakdown

* core.model.data.NewsResource: This is the news resource definition the app sees. This is the specification the entire app builds to. It is exposed from the data module because the entire app needs to know about it.
*  core.network.model.NetworkNewsResource: This is how the app gets news resources from the network. The rest of the app is unaware of this type, only the data layer knows or cares about this. It is fully owned by the data module.
* core.database.model.NewsResourceEntity: This is how the news resources are stored for offline first access. The rest of the app is unaware of this type, only the data layer knows or cares about this. It is fully owned by the data module.
* FollowableTopic: This is a model that has had business logic applied to it. Ordinarily a `Topic` is just a topic. The intent of a user to follow it or not, adds a dimension to it that is out of the scope of a simple topic. A new model is exposed to represent the application of this logic. Crucially, if we decide to move the persistence of following a topic from `DataStore` to `Room`, this model definition would not change. I exists to encapsulate the business logic of following a topic.

## Where to put what

`asEntity`: converts from the network model to the persisted model. There is no other use case for this. Keep it with the network classes and make it internal. No other module should needs to use this method; guard access to it as such.

`core.data.model`: Should be used for models that define the app; they are the contract the app adheres to. They should have no annotations for deserialization that involves the network or saving to the database, or any data layer specific concern.. The UI layer may depend on them, domain layers may depend on them. They are the public API for the data layer.

For more info please checkout the [offline-first guide](https://developer.android.com/topic/architecture/data-layer/offline-first) and the [modularization guide](https://developer.android.com/topic/modularization).
